src/utils/retry.ts

```diff
@@ -0,0 +1,80 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/example-shopping
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {HttpErrors} from '@loopback/rest';
+import {promisify} from 'util';
+import * as debugFactory from 'debug';
+const debug = debugFactory('loopback:example:shopping');
+
+/**
+ * A task that can be retried
+ */
+export interface Task<T> {
+  run(): Promise<T | null | undefined>;
+  // tslint:disable-next-line:no-any
+  isDone?(err: any, result: T | null | undefined): result is T;
+  description: string;
+}
+/**
+ * Retry a task for number of times with the given interval in ms
+ * @param task Task object {run, description}
+ * @param maxRetries Maximum number of tries (including the first run),
+ * default to 10
+ * @param interval Milliseconds to wait after each try, default to 100ms
+ */
+export async function retry<T>(
+  task: Task<T>,
+  maxRetries: number = 10,
+  interval: number = 100,
+): Promise<T> {
+  if (maxRetries < 1) maxRetries = 1;
+  let triesLeft = maxRetries;
+  while (true) {
+    debug(
+      'Try %s (%d/%d)',
+      task.description,
+      maxRetries - triesLeft + 1,
+      maxRetries,
+    );
+    let result;
+    let error = undefined;
+    try {
+      result = await task.run();
+    } catch (err) {
+      debug('Error thrown from task %s: ', task.description, err);
+      error = err;
+    }
+    if (
+      (task.isDone && task.isDone(error, result)) ||
+      taskFinished(error, result)
+    ) {
+      if (error) throw error;
+      return result;
+    }
+    if (--triesLeft > 0) {
+      debug('Wait for %d ms', interval);
+      await sleep(interval);
+    } else {
+      // No more retries, timeout
+      const msg = `Failed to ${task.description} after ${maxRetries *
+        interval} ms`;
+      debug('%s', msg);
+      throw new HttpErrors.RequestTimeout(msg);
+    }
+  }
+}
+
+// tslint:disable-next-line:no-any
+function taskFinished<T>(err: any, result: T | null | undefined): result is T {
+  return err == null && result != null;
+}
+
+/**
+ * Sleep for the given milliseconds
+ * @param ms Number of milliseconds to wait
+ */
+export function sleep(ms: number) {
+  return promisify(setTimeout)(ms);
```

Comment:
Please move `promisify(setTimeout)` out of this function, so that we don't have to create a new function whenever sleep is called.

```ts
const setTimeoutAsync = promisify(setTimeout);

export function sleep(ms) {
  return setTimeoutAsync(ms);
}
```

I think this can be further simplified as follows:

```ts
export const sleep = promisify(setTimeout);
```